### ISSUE TITLE
Middlewares Pattern

### ISSUE THREAD

**Issue by Abiti-936** [httpland/http-router#4]
I liked these Routing Library & I am now using it with My Supabase Edge Functions, & it's just a great library.

One, thing that is cool to have is middlewares pattern, to Add CORS Header, Check for Authorization Headers & more... before & after Handling Routes.

For Example, The Pattern can be like these :- 

```ts
import { createRouter } from "https://deno.land/x/http_router/mod.ts";
import { serve } from "https://deno.land/std/http/mod.ts";
import cors from 'https://deno.land/x/edge_cors/src/cors.ts';

const router = createRouter({
  "/*": async (request, { next }) => {
    const response = await next(); // invokes the next route

    return cors(request, response); // set CORS Headers to the response
  },
  "/hello": (request, _context) => {
    return new Response.json({ hello: "world" });
  }
});

await serve(router);
```

I think these can make a whole lot of use cases possible. Thanks.

**Comment by TomokiMiyauci** [comment]
@Abiti-936 
We are currently looking into supporting nested routes.

One idea is to compose a `createRouter`.
For example, we aim to be able to do the following:

```ts
import cors from "https://deno.land/x/edge_cors/src/cors.ts";
import { createRouter } from "https://deno.land/x/http_router/mod.ts";

const api = createRouter({
  "/hello": () => new Response("hello"),
});
const router = createRouter({
  "/api": cors(api),
});
function log(handler: (req: Request) => Promise<Response> | Response) {
  return async (req: Request) => {
    const res = await handler(req);
    console.log({
      req, res,
    });
    return res;
  };
}
await Deno.serve(log(router));
````

If the router is self-recursive, it is divisible and most expressive and beautiful.

However, it requires some ingenuity in implementation because of the need to perform URL pattern matching.

---

The middleware pattern is difficult to formulate a standard API.

In fact, middleware was discussed in Deno, but is not currently implemented for the same reason. https://github.com/denoland/deno_std/issues/1295

Also, `createRouter` aims to be a stateless function.
Therefore, we do not plan to add methods such as `use`.

Any thoughts on these?

**Comment by Abiti-936** [comment]
I mean I don't expect you to add the **use** method, but you can add middlewares pattern like /*, /api/*, or something like that. Personally I don't like the use method that we can find in [Oak](https://github.com/oakserver/oak#basic-usage) & similar frameworks, but I like the [Netlify Edge Functions Middlewares Pattern](https://github.com/netlify/edge-functions-examples/blob/main/netlify.toml#L12).

Also the Upper Implementation is Just Really Good.

**Comment by Abiti-936** [comment]
I mean I don't expect you to add the **use** method, but you can add middlewares pattern like /*, /api/*, or something like that. Personally I don't like the use method that we can find in [Oak](https://github.com/oakserver/oak#basic-usage) & similar frameworks, but I like the [Netlify Edge Functions Middlewares Pattern](https://github.com/netlify/edge-functions-examples/blob/main/netlify.toml#L12).

Also the Upper Implementation is Just Really Good.

**Comment by Abiti-936** [comment]
I have tried the Upper log example by changing it cors & it's working.

```ts
import { createRouter } from "https://deno.land/x/http_router/mod.ts"; 
import edgeCors from "https://deno.land/x/edge_cors@0.2.1/src/cors.ts"; 
 import { options } from "./corsOptions.ts";

const api = createRouter({ "/hello": () => new Response("hello")}); 
 export function cors(handler: (request: Request) => Promise<Response> | Response) { 
   return async (request: Request) => { 
     const response = await handler(request); 
  
     return edgeCors(request, response, options); 
   }; 
 }

await Deno.serve(cors(router));
```

**Comment by TomokiMiyauci** [comment]
@Abiti-936
I have released v1.2.

A number of important features have been added, including nested route and base path.

https://github.com/httpland/http-router/releases/tag/1.2.0

**Comment by Abiti-936** [comment]
That's Great. I will take a look at the new release. Thanks.
